src/frontend/Typechecker.ml

```diff
@@ -730,34 +730,73 @@ let verify_assignment_global loc cf block is_global id =
+let verify_assignment_operator loc assop lhs rhs =
   let err op =
-    Semantic_error.illtyped_assignment loc op lhs.emeta.type_ rhs.emeta.type_
+    Semantic_error.illtyped_assignment loc op lhs.lmeta.type_ rhs.emeta.type_
   in
-  let () =
-    match assop with
-    | Assign | ArrowAssign ->
-        if
-          UnsizedType.check_of_same_type_mod_array_conv "" lhs.emeta.type_
-            rhs.emeta.type_
-        then ()
-        else err Operator.Equals |> error
-    | OperatorAssign op -> (
-        let args = List.map ~f:arg_type [lhs; rhs] in
-        let return_type =
-          Stan_math_signatures.assignmentoperator_stan_math_return_type op args
-        in
-        match return_type with Some Void -> () | _ -> err op |> error ) in
-  mk_typed_statement ~return_type:NoReturnType ~loc
-    ~stmt:(mk_assignment_from_indexed_expr assop lhs rhs)
+  match assop with
+  | Assign | ArrowAssign ->
+      if
+        UnsizedType.check_of_same_type_mod_array_conv "" lhs.lmeta.type_
+          rhs.emeta.type_
+      then ()
+      else err Operator.Equals |> error
+  | OperatorAssign op -> (
+      let args = List.map ~f:arg_type [Ast.expr_of_lvalue lhs; rhs] in
+      let return_type =
+        Stan_math_signatures.assignmentoperator_stan_math_return_type op args
+      in
+      match return_type with Some Void -> () | _ -> err op |> error )
+
+let check_lvalue cf tenv = function
+  | {lval= LVariable id; lmeta= ({loc} : located_meta)} ->
+      verify_identifier id ;
+      let ad_level, type_ = check_id cf loc tenv id in
+      {lval= LVariable id; lmeta= {ad_level; type_; loc}}
+  | {lval= LIndexed (lval, idcs); lmeta= {loc}} ->
+      let rec check_inner = function
+        | {lval= LVariable id; lmeta= ({loc} : located_meta)} ->
+            verify_identifier id ;
+            let ad_level, type_ = check_id cf loc tenv id in
+            let var = {lval= LVariable id; lmeta= {ad_level; type_; loc}} in
+            (var, var, [])
+        | {lval= LIndexed (lval, idcs); lmeta= {loc}} ->
+            let lval, var, flat = check_inner lval in
+            let idcs = List.map ~f:(check_index cf tenv) idcs in
+            let ad_level =
+              inferred_ad_type_of_indexed lval.lmeta.ad_level idcs in
+            let type_ =
+              inferred_unsizedtype_of_indexed ~loc lval.lmeta.type_ idcs in
+            ( {lval= LIndexed (lval, idcs); lmeta= {ad_level; type_; loc}}
+            , var
+            , flat @ idcs ) in
+      let lval, var, flat = check_inner lval in
+      let idcs = List.map ~f:(check_index cf tenv) idcs in
+      let ad_level = inferred_ad_type_of_indexed lval.lmeta.ad_level idcs in
+      let type_ = inferred_unsizedtype_of_indexed ~loc lval.lmeta.type_ idcs in
+      if List.exists ~f:is_multiindex flat then (
+        add_warning loc
+          "Nested multi-indexing on the left hand side of assignment does not \
+           behave the same as nested indexing in expressions. This is \
```

Comment:
I think we should indicate what it does mean/what it can be replaced with (All other deprecation messages mention the canonicalize flag, like `This can be automatically changed using the canonicalize flag for stanc`)